File: LeetCode/Easy/5_valid_parentheses.py


#In object-oriented programming, a class is a blueprint for creating objects (a particular data structure).
class Solution:
    # define a method name isValid within the solution class
    def isValid(self, s: str) -> bool:
        # this list will be used to store all the opening Brackets
        open_bracket = []
        
        # this dic will maps each closing bracket to its corresponding opening bracket 
        bracket_map = {')': '(', '}': '{', ']': '['}
        print('this is open',bracket_map[')'])
        for i in s:
            print(i)
            if len(s) <= 1:
                return False

            # this line check if i is opening brackets 
            if i in bracket_map.values():
                open_bracket.append(i)
                print('after the append', open_bracket)

            # this lines handle the close bracket 
            elif i in bracket_map.keys():
                # if the list open_bracket is not empty and 
                # list items == bracket_map[i] , eg i = ] |
                #  bracket_map[']'] = [  | we put the keys to retrive the values 
                if open_bracket and open_bracket[-1] == bracket_map[i]:
                    print('thi s', bracket_map[i])
                    # this line remove the last item from open bracket list
                    open_bracket.pop()
                    print('after the pop', open_bracket)

                    #if either open_bracket is empty or the last item in open_bracket does not match the corresponding opening bracket for i
                else:
                    return False
        
        return len(open_bracket) == 0

    
        


s = "([])"

solution = Solution()

result = solution.isValid(s)

print(result)

File: LeetCode/Easy/5_valid_parentheses.ts


function isValid(s: string): boolean {
     //A variable openBracket is initialized as an empty array
     let openBracket: string[] = [];

     const bracketMap: { [key: string]: string } = { ')': '(', '}': '{', ']': '[' };

     for (let i of s) {
         if (s.length <= 1) {
             return false;
         }
         //Check if the Character is an Open Bracket 
         //Object.values(bracketMap) retrieves the values of the bracketMap object, which are the open brackets
         if (Object.values(bracketMap).includes(i)) {
             openBracket.push(i);
     
             //Check if the Character is a Closing Bracket 
             //Object.keys(bracketMap) retrieves the keys of the bracketMap object, which are the closing brackets
         } else if (Object.keys(bracketMap).includes(i)) {

          //openBracket.length - 1 | to get the last element 
             if (openBracket.length && openBracket[openBracket.length - 1] === bracketMap[i]) {
                 openBracket.pop();
             } else {
                 return false;
             }
         }
     }

     return openBracket.length === 0;
 }


// Example usage:
const s = "()[]{";
const solution = isValid(s)
console.log(isValid(s));  // This should print false

